File: core/i18n.py
"""Internationalization (i18n) hook for EnglishCoach Pro.

Provides bilingual labels (English / Vietnamese) for UI elements.
Usage:
    from core.i18n import t
    label = ctk.CTkLabel(text=t("dashboard"))
"""

import json
import os

# Default language — can be overridden by user settings
_current_lang = "en"

# Built-in translations
_TRANSLATIONS = {
    "en": {
        "app_title": "EnglishCoach Pro",
        "dashboard": "Dashboard",
        "vocabulary": "Vocabulary",
        "grammar": "Grammar",
        "reading": "Reading",
        "listening": "Listening",
        "writing": "Writing",
        "speaking": "Speaking",
        "mock_test": "Mock Test",
        "progress": "Progress",
        "planner": "Planner",
        "settings": "Settings",
        "placement": "Placement Test",
        "onboarding": "Getting Started",
        "welcome_back": "Welcome back",
        "days_to_exam": "days to exam",
        "streak": "Streak",
        "daily_goal": "Daily Goal",
        "todays_plan": "Today's Plan",
        "word_of_the_day": "Word of the Day",
        "words_learned": "Words Learned",
        "current_band": "Current Band",
        "xp_this_week": "XP This Week",
        "start_test": "Start Test",
        "submit": "Submit",
        "next": "Next",
        "back": "Back",
        "finish": "Finish",
        "save": "Save",
        "cancel": "Cancel",
        "correct": "Correct",
        "incorrect": "Incorrect",
        "again": "Again",
        "hard": "Hard",
        "good": "Good",
        "easy": "Easy",
        "show_meaning": "Click card to reveal meaning",
        "play_audio": "Play Audio",
        "get_feedback": "Get Feedback",
        "evaluate": "Evaluate",
        "generate_plan": "Generate Study Plan",
        "review_weak_areas": "Review Weak Areas",
        "error_journal": "Error Journal",
        "activity_heatmap": "Activity Heatmap",
        "skill_balance": "Skill Balance",
        "band_over_time": "Band Estimate Over Time",
        "time_per_skill": "Time Spent per Skill",
    },
    "vi": {
        "app_title": "EnglishCoach Pro",
        "dashboard": "Bảng điều khiển",
        "vocabulary": "Từ vựng",
        "grammar": "Ngữ pháp",
        "reading": "Đọc hiểu",
        "listening": "Nghe hiểu",
        "writing": "Viết",
        "speaking": "Nói",
        "mock_test": "Thi thử",
        "progress": "Tiến độ",
        "planner": "Lịch học",
        "settings": "Cài đặt",
        "placement": "Bài kiểm tra xếp hạng",
        "onboarding": "Bắt đầu",
        "welcome_back": "Chào mừng trở lại",
        "days_to_exam": "ngày đến kỳ thi",
        "streak": "Chuỗi ngày",
        "daily_goal": "Mục tiêu hàng ngày",
        "todays_plan": "Kế hoạch hôm nay",
        "word_of_the_day": "Từ vựng hôm nay",
        "words_learned": "Từ đã học",
        "current_band": "Trình độ hiện tại",
        "xp_this_week": "XP tuần này",
        "start_test": "Bắt đầu",
        "submit": "Nộp bài",
        "next": "Tiếp",
        "back": "Quay lại",
        "finish": "Hoàn thành",
        "save": "Lưu",
        "cancel": "Hủy",
        "correct": "Đúng",
        "incorrect": "Sai",
        "again": "Lặp lại",
        "hard": "Khó",
        "good": "Tốt",
        "easy": "Dễ",
        "show_meaning": "Nhấp thẻ để xem nghĩa",
        "play_audio": "Phát âm thanh",
        "get_feedback": "Nhận phản hồi",
        "evaluate": "Đánh giá",
        "generate_plan": "Tạo kế hoạch học",
        "review_weak_areas": "Ôn tập điểm yếu",
        "error_journal": "Nhật ký lỗi",
        "activity_heatmap": "Bản đồ hoạt động",
        "skill_balance": "Cân bằng kỹ năng",
        "band_over_time": "Tiến độ trình độ",
        "time_per_skill": "Thời gian theo kỹ năng",
    },
}


def set_language(lang: str):
    """Set the current language ('en' or 'vi')."""
    global _current_lang
    if lang in _TRANSLATIONS:
        _current_lang = lang


def get_language() -> str:
    return _current_lang


def t(key: str, lang: str = None) -> str:
    """Translate a key to the current (or specified) language.

    Falls back to English if the key is not found in the target language,
    and to the key itself if not found at all.
    """
    use_lang = lang or _current_lang
    translations = _TRANSLATIONS.get(use_lang, _TRANSLATIONS["en"])
    return translations.get(key, _TRANSLATIONS["en"].get(key, key))


def available_languages() -> dict:
    """Return available languages as {code: display_name}."""
    return {"en": "English", "vi": "Tiếng Việt"}

File: web/src/lib/i18n.ts
import i18n from 'i18next'
import { initReactI18next } from 'react-i18next'
import { useSettingsStore } from '@/stores/settings.store'

const resources = {
  en: {
    translation: {
      // ── Nav ──
      'nav.home': 'Home',
      'nav.vocab': 'Vocab',
      'nav.grammar': 'Grammar',
      'nav.listening': 'Listening',
      'nav.reading': 'Reading',
      'nav.writing': 'Writing',
      'nav.speaking': 'Speaking',
      'nav.mock': 'Mock',
      'nav.progress': 'Progress',
      'nav.settings': 'Settings',

      // ── Dashboard ──
      'dashboard.greeting.morning': 'Good morning',
      'dashboard.greeting.afternoon': 'Good afternoon',
      'dashboard.greeting.evening': 'Good evening',
      'dashboard.daysToExam': 'days to exam',
      'dashboard.dayStreak': 'day streak',
      'dashboard.dailyGoal': 'Daily Goal',
      'dashboard.wordsLearned': 'Words Learned',
      'dashboard.currentBand': 'Current Band',
      'dashboard.xpThisWeek': 'XP This Week',
      'dashboard.todaysPlan': "Today's Plan",
      'dashboard.wordOfDay': 'Word of the Day',
      'dashboard.noPlan': 'No plan for today yet.',
      'dashboard.generatePlan': 'Generate a study plan →',

      // ── Vocabulary ──
      'vocab.title': 'Vocabulary Practice',
      'vocab.tapToReveal': 'Tap to reveal',
      'vocab.again': 'Again',
      'vocab.hard': 'Hard',
      'vocab.good': 'Good',
      'vocab.easy': 'Easy',
      'vocab.reviewed': 'Cards reviewed',
      'vocab.accuracy': 'Accuracy',
      'vocab.xpEarned': 'XP earned',
      'vocab.nextSession': 'Next session',
      'vocab.tomorrow': 'Tomorrow',
      'vocab.noCards': 'No cards due. Come back tomorrow!',

      // ── Listening ──
      'listening.title': 'Listening Practice',
      'listening.playAudio': 'Play Audio',
      'listening.loading': 'Loading listening tests…',
      'listening.noTests': 'No listening tests available.',
      'listening.results': 'Listening Test Results',
      'listening.estimatedBand': 'Estimated IELTS Band',
      'listening.score': 'Score',

      // ── Speaking ──
      'speaking.title': 'Speaking Practice',
      'speaking.playExaminer': 'Play Examiner (Correct Pronunciation)',
      'speaking.record': 'Record & Evaluate',
      'speaking.listening': 'Listening…',
      'speaking.accuracy': 'Pronunciation Accuracy',
      'speaking.wordsToPractice': 'Words to practice',
      'speaking.startTimer': 'Start Timer',
      'speaking.startPrep': 'Start 1-minute Preparation Timer',

      // ── Writing ──
      'writing.title': 'Writing Practice',
      'writing.taskPrompt': 'Task Prompt',
      'writing.yourEssay': 'Your Essay',
      'writing.getAiFeedback': 'Get AI Feedback',
      'writing.quickLocal': 'Quick Local Analysis (no server needed)',
      'writing.analyzing': 'Analyzing…',
      'writing.wordsMin': 'words minimum',
      'writing.estimatedBand': 'Estimated Band Score',
      'writing.assessmentCriteria': 'Assessment Criteria',

      // ── Progress ──
      'progress.title': 'Your Progress',
      'progress.overview': 'Overview',
      'progress.charts': 'Charts',
      'progress.errorJournal': 'Error Journal',

      // ── Common ──
      'common.loading': 'Loading…',
      'common.next': 'Next',
      'common.back': 'Back',
      'common.save': 'Save',
      'common.cancel': 'Cancel',
      'common.restart': 'Restart',
      'common.level': 'Level',
      'common.xp': 'XP',
    },
  },
  vi: {
    translation: {
      // ── Nav ──
      'nav.home': 'Trang chủ',
      'nav.vocab': 'Từ vựng',
      'nav.grammar': 'Ngữ pháp',
      'nav.listening': 'Nghe',
      'nav.reading': 'Đọc',
      'nav.writing': 'Viết',
      'nav.speaking': 'Nói',
      'nav.mock': 'Thi thử',
      'nav.progress': 'Tiến độ',
      'nav.settings': 'Cài đặt',

      // ── Dashboard ──
      'dashboard.greeting.morning': 'Chào buổi sáng',
      'dashboard.greeting.afternoon': 'Chào buổi chiều',
      'dashboard.greeting.evening': 'Chào buổi tối',
      'dashboard.daysToExam': 'ngày đến kỳ thi',
      'dashboard.dayStreak': 'ngày liên tục',
      'dashboard.dailyGoal': 'Mục tiêu hàng ngày',
      'dashboard.wordsLearned': 'Từ đã học',
      'dashboard.currentBand': 'Band hiện tại',
      'dashboard.xpThisWeek': 'XP tuần này',
      'dashboard.todaysPlan': 'Kế hoạch hôm nay',
      'dashboard.wordOfDay': 'Từ vựng hôm nay',
      'dashboard.noPlan': 'Chưa có kế hoạch cho hôm nay.',
      'dashboard.generatePlan': 'Tạo kế hoạch học tập →',

      // ── Vocabulary ──
      'vocab.title': 'Luyện tập Từ vựng',
      'vocab.tapToReveal': 'Chạm để xem',
      'vocab.again': 'Lại',
      'vocab.hard': 'Khó',
      'vocab.good': 'Tốt',
      'vocab.easy': 'Dễ',
      'vocab.reviewed': 'Thẻ đã ôn',
      'vocab.accuracy': 'Độ chính xác',
      'vocab.xpEarned': 'XP đạt được',
      'vocab.nextSession': 'Buổi học tiếp theo',
      'vocab.tomorrow': 'Ngày mai',
      'vocab.noCards': 'Không có thẻ cần ôn. Hẹn gặp lại ngày mai!',

      // ── Listening ──
      'listening.title': 'Luyện Nghe',
      'listening.playAudio': 'Phát âm thanh',
      'listening.loading': 'Đang tải bài nghe…',
      'listening.noTests': 'Không có bài nghe nào.',
      'listening.results': 'Kết quả bài nghe',
      'listening.estimatedBand': 'Band IELTS ước tính',
      'listening.score': 'Điểm',

      // ── Speaking ──
      'speaking.title': 'Luyện Nói',
      'speaking.playExaminer': 'Phát âm chuẩn (Người khảo)',
      'speaking.record': 'Ghi âm & Đánh giá',
      'speaking.listening': 'Đang nghe…',
      'speaking.accuracy': 'Độ chính xác phát âm',
      'speaking.wordsToPractice': 'Từ cần luyện thêm',
      'speaking.startTimer': 'Bật đồng hồ',
      'speaking.startPrep': 'Bật 1 phút chuẩn bị',

      // ── Writing ──
      'writing.title': 'Luyện Viết',
      'writing.taskPrompt': 'Đề bài',
      'writing.yourEssay': 'Bài viết của bạn',
      'writing.getAiFeedback': 'Nhận phản hồi AI',
      'writing.quickLocal': 'Phân tích nhanh (không cần server)',
      'writing.analyzing': 'Đang phân tích…',
      'writing.wordsMin': 'từ tối thiểu',
      'writing.estimatedBand': 'Band ước tính',
      'writing.assessmentCriteria': 'Tiêu chí đánh giá',

      // ── Progress ──
      'progress.title': 'Tiến độ của bạn',
      'progress.overview': 'Tổng quan',
      'progress.charts': 'Biểu đồ',
      'progress.errorJournal': 'Nhật ký lỗi',

      // ── Common ──
      'common.loading': 'Đang tải…',
      'common.next': 'Tiếp theo',
      'common.back': 'Quay lại',
      'common.save': 'Lưu',
      'common.cancel': 'Hủy',
      'common.restart': 'Làm lại',
      'common.level': 'Cấp độ',
      'common.xp': 'XP',
    },
  },
}

// Initialize with stored language or default to 'en'
const storedLang = (() => {
  try {
    const raw = localStorage.getItem('settings-store')
    if (raw) {
      const parsed = JSON.parse(raw)
      return parsed?.state?.language || 'en'
    }
  } catch { /* ignore */ }
  return 'en'
})()

i18n
  .use(initReactI18next)
  .init({
    resources,
    lng: storedLang,
    fallbackLng: 'en',
    interpolation: { escapeValue: false },
  })

// Sync i18n language with settings store
useSettingsStore.subscribe((state) => {
  if (state.language && state.language !== i18n.language) {
    i18n.changeLanguage(state.language)
  }
})

export default i18n
